docs/design_notes/setup_or_nncompact.rst

```diff
@@ -0,0 +1,78 @@
+Should I use ``setup`` or ``nn.compact``?
+=========================================
+
+In Flax's module system (named `Linen`_), submodules, parameters and/or other variables
+can be defined in two ways:
+
+1. **Explicitly** (using ``setup``):
+
+   Assign to ``self.<attr>`` inside a :meth:`setup <flax.linen.Module.setup>`
+   method. Then read ``self.<attr>`` from any "forward pass" method defined on the class.
+   This resembles how most objects work in Python, and how modules are defined in PyTorch.
+2. **In-line** (using ``nn.compact``):
+  
+   Declare directly within a single "forward pass" method annotated
+   with :meth:`nn.compact <flax.linen.compact>`. This allows you to define your whole module
+   in a single method, and "co-locate" submodules, parameters and variables next to 
+   where they are used.
+
+**Both of these approaches are perfectly valid, behave the same way, and interoperate with all of Flax**.
+
+Here is a short example of a module defined in both ways. 
+
+.. testsetup::
+
+  import flax.linen as nn
+
+.. codediff:: 
+  :title_left: Using ``setup``
+  :title_right: Using ``nn.compact``
+  
+  class MLP(nn.Module):
+    def setup(self):
+      self.dense1 = nn.Dense(32)
+      self.dense2 = nn.Dense(32)
+
+    def __call__(self, x):
+      x = self.dense1(x)
+      x = nn.relu(x)
+      x = self.dense2(x)
+      return x
+  ---
+  class MLP(nn.Module):
+
+
+
+    @nn.compact
+    def __call__(self, x):
+      x = nn.Dense(32, name="dense1")(x)
+      x = nn.relu(x)
+      x = nn.Dense(32, name="dense2")(x)
+      return x
+
+So, how would you decide which style to use? It can be a matter of taste, but here are some pros and cons:
```

Comment:
Why I see *extremely* often is that users use both `@nn.compact` and `setup()`. It would be useful to write down what the consequence is of that. (To be honest I'd like to know as well!)

Reply:
Yeah good point. It actually works nicely. But maybe that should be in an footnote to this doc rather than in the main text?